### ISSUE TITLE
Parsing with YARD - Sketchup/Entities.rb won't load

### ISSUE THREAD

**Issue by MSP-Greg** [SketchUp/ruby-api-stubs#53]
Thom,

I was just looking thru my docs, and I noticed `Sketchup::Entities` wasn't there.

The following line [`Sketchup/Entities.rb:359`](https://github.com/SketchUp/ruby-api-stubs/blob/860f5f21524c955aa5d24c52c95074a1c0b8ee9f/SketchUp/Sketchup/Entities.rb#L359) stops parsing:

```ruby
def add_edges(point1, point2, ...)
```

Maybe it could be:
```ruby
def add_edges(*points)
```

I guess not a lot of people are ripping their own docs.  I also have to check YARD's code and my code, as I'm not seeing any errors.  Then again, YARD may assume that the code being doc's does load in Ruby...

Hope things are well & Happy Holidays...

Greg

**Comment by thomthom** [comment]
Ah, I see. This is probably due to the fact that we use and `@overload` in our source code. But when we generate stubs, if there is only one `@overload` the `@overload` tag is stripped. And it appear that affects the parsing.

**Comment by MSP-Greg** [comment]
Sorry I wasn't a bit more clear.  This isn't related to comments in the files.  

I think YARD assumes that your set of files all work correctly and load.  `Sketchup/Entities.rb` won't load with `require`/`require_relative`, etc.

But, the good news is that it appears that file is the only one that does not.

Below is output, showing loading errors:
```
E:\>ruby --disable-gems E:\GitHub\ruby-api-stubs\ruby_api_stubs_check.rb

Layout/AngularDimension.rb              uninitialized constant Layout::Entity
Layout/Ellipse.rb                       uninitialized constant Layout::Entity
Layout/FormattedText.rb                 uninitialized constant Entity
Layout/LockedEntityError.rb             uninitialized constant ArgError
Layout/LockedLayerError.rb              uninitialized constant ArgError
Sketchup/ArcCurve.rb                    uninitialized constant Sketchup::Curve
Sketchup/DefinitionObserver.rb          uninitialized constant Sketchup::EntityObserver
Sketchup/Entities.rb                    E:/GitHub/ruby-api-stubs/SketchUp/Sketchup/Entities.rb:359: syntax error, unexpected ...
...f add_edges(point1, point2, ...)
...                            ^~~
Sketchup/Http/Request.rb                uninitialized constant Sketchup::Http
Sketchup/Http/Response.rb               uninitialized constant Sketchup::Http
Sketchup/Licensing/ExtensionLicense.rb  uninitialized constant Sketchup::Licensing
----------------------------------------
Checked 142 files of 142
```

For code like this to work, there there are two issues:
1. Namespacing should be declared as:
```ruby
module SketchUp
class Entities
end
end
```
or:
```ruby
module SketchUp ; class Entities
end ; end
```
not:
```ruby
class SketchUp::Entities
end
```

2. Parent classes - any items that are parent classes need to be 'autoloaded'.

Below is the code, add to the root of your repo:
```ruby
# frozen_string_literal: true

# Load these to avoid some namespacing issues
module Geom     ; end
module Layout   ; end
module Sketchup ; end
module UI       ; end

module RubyAPIStubsCheck

  WID = 40

  require_relative 'SketchUp/sketchup'
  require_relative 'SketchUp/Sketchup/Entity'
  require_relative 'SketchUp/Sketchup/DrawingElement'
  
  def self.run
    Dir.chdir("#{__dir__}/SketchUp") do |d|
      ary = Dir.glob('**/*.rb')
      cntr = 0
      ary.each { |fn|
        check_require fn
        cntr += 1
      }
      puts "#{'-' * WID}\nChecked #{cntr} files of #{ary.length}"
    end
  end

  def self.check_require(fn)
    require_relative "SketchUp/#{fn}"
  rescue LoadError => e
    puts "#{fn.ljust(WID)}#{e.message}"
  rescue NameError => e
    puts "#{fn.ljust(WID)}#{e.message}"
  rescue SyntaxError => e
    puts "#{fn.ljust(WID)}#{e.message}"
  end
  
end
RubyAPIStubsCheck.run
```

Finally, I do intend to get back to the UI Dialog framework I talked about.  Been busy with some ruby related things.  Have a good new year...  Greg

**Comment by MSP-Greg** [comment]
See PR #55 & #56

**Comment by DanRathbun** [comment]
> @MSP-Greg This isn't related to comments in the files.

**FYI**

Thomas is referring to the **comments** in the SketchUp **core** code that describe the Ruby API object being defined on the C side. The [custom YARD template ](https://github.com/SketchUp/sketchup-yard-template) is run on the **core C code** to produce the [stubs](https://github.com/SketchUp/ruby-api-stubs) and the [docs](https://github.com/SketchUp/ruby-api-docs) repos.

The **stubs** repo has 2 _**primary**_ duties. To provide autocompletion for IDEs, and to accept edits that are "sucked" back into the core C code updating the inline API "docu-comments".

Trimble does not generate the **docs** repo directly from the **stubs** repo (like us mere mortals do,) they run YARD on the SketchUp C core code to output the **docs** repo, that serves the website.

The SketchUp team did not always use YARD. Before this (in the Google ownership years,) they used something custom similar to RDoc to read the C code and produce the old documentation website.
Sometimes old syntax (or style) in the "docu-comments" causes funny errors in the new YARD docs.

When Thomas says that `@overload` tags are stripped if there is only 1,... he means the [custom YARD template](https://github.com/SketchUp/sketchup-yard-template) that he uses to write out the **stubs**, is the culprit.

So, it is either fix the C-side "docu-comments" or fix the template that is causing the booboo, or both.

**Comment by MSP-Greg** [comment]
> Trimble does not generate the docs repo directly from the stubs repo (like us mere mortals do,) they run YARD on the SketchUp C core code to output the docs repo, that serves the website.

I'm aware of that.  Remember, I have hacked newer versions of ruby into SU, so I've got a pretty good idea where the code is...

> When Thomas says that @overload tags are stripped if there is only 1,... he means the custom YARD template that he uses to write out the stubs, is the culprit.

I think it's just the overload tag content, not an issue with the template.  Both YARD & RDoc have syntaxes for defining Ruby method signatures in c source for documentation, and as I recall, neither require them to be 'parsable', which is causing the issue with the stub code.

Thanks, Greg

**Comment by MSP-Greg** [comment]
@DanRathbun or @thomthom

I assume that changes to this repo are manually moved to the c source?

**Comment by DanRathbun** [comment]
NO Thomas has a script or a special YARD template, that reads the updated stubs and updates the C source comments. He has mentioned how it works in "issue talk", here or in the template repo before.

**It is definitely not manual.**

**Comment by thomthom** [comment]
Yes, I have a script that generate stubs from the C++ source. And where we some times have single `@overload` tags in the original source I reformat things in the stubs. If there is only one `@overload` signature it get's converted into a method signature in the stubs.

For instance, optional arguments must be described with `@overload` in C++ source, as it's not possible to infer them from the C++ function signature.

So while a method signature is in comment only in the C++ source it might end up as begin the signature in the Ruby source.

I also had a script that merged changes in the stubs back into our C++ source. But I'm not sure if it's working any more after SU2018. We had to move things around when we added the LayOut Ruby API.
